<a href="https://colab.research.google.com/github/eridanletalis/GB_NLP/blob/main/hw_lesson_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
import numpy as np
import re
import matplotlib.pyplot as plt
import seaborn as sns
import nltk
import warnings 
warnings.filterwarnings("ignore", category=DeprecationWarning)
import os

In [4]:
# стандартным методом через pd.pickle_read() открываться не хотел

In [5]:
!pip3 install pickle5

import pickle5 as pickle

In [6]:
with open("./preprocessed_ds.pkl", "rb") as fh:
  data = pd.DataFrame(pickle.load(fh))

In [7]:
data

id  ...                                   tweet_lemmatized
0          1  ...  father dysfunctional selfish drag kid dysfunct...
1          2  ...  thank lyft credit use cause offer wheelchair v...
2          3  ...                                     bihday majesty
3          4  ...                            model love take time ur
4          5  ...                      factsguide society motivation
...      ...  ...                                                ...
49154  49155  ...  think factory leave right polarisation trump u...
49155  49156  ...  feel like mermaid hairflip neverready formal w...
49156  49157  ...  hillary campaign today ohio omg amp use word l...
49157  49158  ...  happy work conference right mindset lead cultu...
49158  49159  ...  song glad free download shoegaze newmusic newsong

[49159 rows x 7 columns]

In [8]:
# По результатам предыдущего дз в нужных столбцах уже отсутствуют стоп-слова

In [9]:
tweet_stemmed_doc = list(data['tweet_stemmed'])
# tweet_stemmed_doc

In [10]:
tweet_lemmatized_doc = list(data['tweet_lemmatized'])
# tweet_lemmatized_doc

Игнорируем слова, частота которых в документе строго превышает порог 0.9 с помощью max_df.
Ограничим количество слов, попадающий в мешок, с помощью max_features = 1000.


In [11]:
from sklearn.feature_extraction.text import CountVectorizer

In [12]:
stemmed_count_vectorizer = CountVectorizer(ngram_range=(1, 1), 
                                   analyzer='word', 
                                   binary=False, 
                                   tokenizer=str.split,
                                   max_df=0.9,
                                   max_features=1000,
                                   stop_words='english')

In [13]:
lemmatized_count_vectorizer = CountVectorizer(ngram_range=(1, 1), 
                                   analyzer='word', 
                                   binary=False, 
                                   tokenizer=str.split,
                                   max_df=0.9,
                                   max_features=1000,
                                   stop_words='english')

In [14]:
stemmed_bow = stemmed_count_vectorizer.fit_transform(tweet_stemmed_doc)

In [15]:
lemmatized_bow = lemmatized_count_vectorizer.fit_transform(tweet_lemmatized_doc)

In [16]:
feature_names_s = stemmed_count_vectorizer.get_feature_names()
pd.DataFrame(stemmed_bow.toarray(), columns = feature_names_s)

abl  absolut  accept  account  act  ...  york  young  youtub  yr  yummi
0        0        0       0        0    0  ...     0      0       0   0      0
1        0        0       0        0    0  ...     0      0       0   0      0
2        0        0       0        0    0  ...     0      0       0   0      0
3        0        0       0        0    0  ...     0      0       0   0      0
4        0        0       0        0    0  ...     0      0       0   0      0
...    ...      ...     ...      ...  ...  ...   ...    ...     ...  ..    ...
49154    0        0       0        0    0  ...     0      0       0   0      0
49155    0        0       0        0    0  ...     0      0       0   0      0
49156    0        0       0        0    0  ...     0      0       0   0      0
49157    0        0       0        0    0  ...     0      0       0   0      0
49158    0        0       0        0    0  ...     0      0       0   0      0

[49159 rows x 1000 columns]

In [17]:
feature_names_l = lemmatized_count_vectorizer.get_feature_names()
pd.DataFrame(lemmatized_bow.toarray(), columns = feature_names_l)

able  absolutely  accept  account  act  ...  youth  youtube  yr  yrs  yummy
0         0           0       0        0    0  ...      0        0   0    0      0
1         0           0       0        0    0  ...      0        0   0    0      0
2         0           0       0        0    0  ...      0        0   0    0      0
3         0           0       0        0    0  ...      0        0   0    0      0
4         0           0       0        0    0  ...      0        0   0    0      0
...     ...         ...     ...      ...  ...  ...    ...      ...  ..  ...    ...
49154     0           0       0        0    0  ...      0        0   0    0      0
49155     0           0       0        0    0  ...      0        0   0    0      0
49156     0           0       0        0    0  ...      0        0   0    0      0
49157     0           0       0        0    0  ...      0        0   0    0      0
49158     0           0       0        0    0  ...      0        0   0    0      0

[49159 rows x 1000 columns]

In [18]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

In [19]:
# TfidfVectorizer?

In [20]:
# tfidf_vectorizer = TfidfVectorizer()

In [21]:
stemmed_value = TfidfVectorizer(ngram_range=(1, 1), 
                                   analyzer='word', 
                                   binary=False, 
                                   tokenizer=str.split,
                                   max_df=0.9,
                                   max_features=1000,
                                   stop_words='english')

In [22]:
lemmatized_value = TfidfVectorizer(ngram_range=(1, 1), 
                                   analyzer='word', 
                                   binary=False, 
                                   tokenizer=str.split,
                                   max_df=0.9,
                                   max_features=1000,
                                   stop_words='english')

In [23]:
stemmed_bow_tfidf = stemmed_value.fit_transform(tweet_stemmed_doc)

In [24]:
lemmatized_bow_tfidf = lemmatized_value.fit_transform(tweet_lemmatized_doc)

In [25]:
feature_names_s_tfidf = stemmed_value.get_feature_names()
pd.DataFrame(stemmed_bow_tfidf.toarray(), columns = feature_names_s_tfidf)

abl  absolut  accept  account  act  ...  york  young  youtub   yr  yummi
0      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
1      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
2      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
3      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
4      0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
...    ...      ...     ...      ...  ...  ...   ...    ...     ...  ...    ...
49154  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
49155  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
49156  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
49157  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0
49158  0.0      0.0     0.0      0.0  0.0  ...   0.0    0.0     0.0  0.0    0.0

[49159 rows x 1000 columns]

In [26]:
feature_names_l_tfidf = lemmatized_value.get_feature_names()
pd.DataFrame(lemmatized_bow_tfidf.toarray(), columns = feature_names_l_tfidf)

able  absolutely  accept  account  act  ...  youth  youtube   yr  yrs  yummy
0       0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
1       0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
2       0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
3       0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
4       0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
...     ...         ...     ...      ...  ...  ...    ...      ...  ...  ...    ...
49154   0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
49155   0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
49156   0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
49157   0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0
49158   0.0         0.0     0.0      0.0  0.0  ...    0.0      0.0  0.0  0.0    0.0

[49159 rows x 1000 columns]

In [27]:
data = open('./corpus.txt').read()
labels, texts = [], []
for i, line in enumerate(data.split("\n")):
    content = line.split()
    labels.append(content[0])
    texts.append(" ".join(content[1:]))

In [28]:
# создаем df
trainDF = pd.DataFrame()
trainDF['text'] = texts
trainDF['label'] = labels
trainDF.head(5)

text       label
0  Stuning even for the non-gamer: This sound tra...  __label__2
1  The best soundtrack ever to anything.: I'm rea...  __label__2
2  Amazing!: This soundtrack is my favorite music...  __label__2
3  Excellent Soundtrack: I truly like this soundt...  __label__2
4  Remember, Pull Your Jaw Off The Floor After He...  __label__2

In [29]:
from sklearn import model_selection, preprocessing, linear_model

train_x, valid_x, train_y, valid_y = model_selection.train_test_split(trainDF['text'], trainDF['label'])

# labelEncode целевую переменную
encoder = preprocessing.LabelEncoder()
train_y = encoder.fit_transform(train_y)
valid_y = encoder.fit_transform(valid_y)

In [30]:
stemmed_bow_fit =  stemmed_value.fit(tweet_stemmed_doc)

/usr/local/lib/python3.7/dist-packages/sklearn/feature_extraction/text.py:507: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn("The parameter 'token_pattern' will not be used"


In [31]:
xtrain_count =  stemmed_bow_fit.transform(train_x)
xvalid_count =  stemmed_bow_fit.transform(valid_x)

classifier = linear_model.LogisticRegression()
classifier.fit(xtrain_count, train_y)
predictions = classifier.predict(xvalid_count)
predictions

array([1, 0, 0, ..., 0, 1, 1])

In [32]:
from sklearn.metrics import accuracy_score

In [33]:
accuracy_score(valid_y, predictions)

0.7208

#Эксперимент

In [34]:
max_featureses = [10, 100, 1000, 2000, 5000, 10000]
max_dfs = [0.5, 0.6, 0.7, 0.8, 0.9, 0.95, 0.99]
ngram_ranges = [(x,y) for x in range(1,4) for y in range(1,4) if x <= y]
ngram_ranges

[(1, 1), (1, 2), (1, 3), (2, 2), (2, 3), (3, 3)]

In [40]:
results_stemmed_count = np.zeros((len(max_featureses), 
                                  len(max_dfs), 
                                  len(ngram_ranges)))

results_lemmatized_count = np.zeros((len(max_featureses), 
                                  len(max_dfs), 
                                  len(ngram_ranges)))

results_stemmed_tfidf = np.zeros((len(max_featureses), 
                                  len(max_dfs), 
                                  len(ngram_ranges)))

results_lemmatized_tfidf = np.zeros((len(max_featureses), 
                                  len(max_dfs), 
                                  len(ngram_ranges)))

In [36]:
warnings.filterwarnings("ignore")

In [ ]:
for i, max_features in enumerate(max_featureses):
  for j, max_df in enumerate(max_dfs):
    for k, ngram_range in enumerate(ngram_ranges):

      # Stemmed and count vectorization
      stemmed_count_vectorizer = CountVectorizer(ngram_range=ngram_range, 
                                   analyzer='word', 
                                   binary=False, 
                                   tokenizer=str.split,
                                   max_df=max_df,
                                   max_features=max_features,
                                   stop_words='english')
      stemmed_bow_fit = stemmed_count_vectorizer.fit(tweet_stemmed_doc)
      xtrain_count =  stemmed_bow_fit.transform(train_x)
      xvalid_count =  stemmed_bow_fit.transform(valid_x)
      classifier = linear_model.LogisticRegression()
      classifier.fit(xtrain_count, train_y)
      predictions = classifier.predict(xvalid_count)
      results_stemmed_count[i,j,k] = accuracy_score(valid_y, predictions)

      # lemmatized and count vectorization
      lemmatized_count_vectorizer = CountVectorizer(ngram_range=ngram_range, 
                                   analyzer='word', 
                                   binary=False, 
                                   tokenizer=str.split,
                                   max_df=max_df,
                                   max_features=max_features,
                                   stop_words='english')
      lemmatized_bow_fit = lemmatized_count_vectorizer.fit(tweet_lemmatized_doc)
      xtrain_count =  lemmatized_bow_fit.transform(train_x)
      xvalid_count =  lemmatized_bow_fit.transform(valid_x)
      classifier = linear_model.LogisticRegression()
      classifier.fit(xtrain_count, train_y)
      predictions = classifier.predict(xvalid_count)
      results_lemmatized_count[i,j,k] = accuracy_score(valid_y, predictions)

      # Stemmed and tfidf
      stemmed_tfidf = TfidfVectorizer(ngram_range=ngram_range, 
                                   analyzer='word', 
                                   binary=False, 
                                   tokenizer=str.split,
                                   max_df=max_df,
                                   max_features=max_features,
                                   stop_words='english')
      stemmed_bow_fit_tfidf = stemmed_tfidf.fit(tweet_stemmed_doc)
      xtrain_count =  stemmed_bow_fit_tfidf.transform(train_x)
      xvalid_count =  stemmed_bow_fit_tfidf.transform(valid_x)
      classifier = linear_model.LogisticRegression()
      classifier.fit(xtrain_count, train_y)
      predictions = classifier.predict(xvalid_count)
      results_stemmed_tfidf[i,j,k] = accuracy_score(valid_y, predictions)

      # lemmatized and tfidf
      lemmatized_tfidf = TfidfVectorizer(ngram_range=ngram_range, 
                                   analyzer='word', 
                                   binary=False, 
                                   tokenizer=str.split,
                                   max_df=max_df,
                                   max_features=max_features,
                                   stop_words='english')
      lemmatized_tfidf = lemmatized_tfidf.fit(tweet_lemmatized_doc)
      xtrain_count =  lemmatized_tfidf.transform(train_x)
      xvalid_count =  lemmatized_tfidf.transform(valid_x)
      classifier = linear_model.LogisticRegression()
      classifier.fit(xtrain_count, train_y)
      predictions = classifier.predict(xvalid_count)
      results_lemmatized_tfidf[i,j,k] = accuracy_score(valid_y, predictions)



In [76]:
# Посмотрим на параметры максимального скора для каждой категории для каждого 
# параметра для n-грамм

print("Максимальный скор каждого набора:")

print(f"results_stemmed_count {results_stemmed_count.max()}")
print(f"results_lemmatized_count {results_lemmatized_count.max()}")
print(f"results_stemmed_tfidf {results_stemmed_tfidf.max()}")
print(f"results_lemmatized_tfidf {results_lemmatized_tfidf.max()}")


for i, ngram_range in enumerate(ngram_ranges):
  print(f"results_stemmed_count для набора {ngram_range}")
  print(f"максимальный скор: {results_stemmed_count[:,:,i].max()}")
  print(results_stemmed_count[:,:,i])

  print(f"results_lemmatized_count для набора {ngram_range}")
  print(f"максимальный скор: {results_lemmatized_count[:,:,i].max()}")
  print(results_lemmatized_count[:,:,i])

  print(f"results_stemmed_tfidf для набора {ngram_range}")
  print(f"максимальный скор: {results_stemmed_tfidf[:,:,i].max()}")
  print(results_stemmed_tfidf[:,:,i])

  print(f"results_lemmatized_tfidf для набора {ngram_range}")
  print(f"максимальный скор: {results_lemmatized_tfidf[:,:,i].max()}")
  print(results_lemmatized_tfidf[:,:,i])



Максимальный скор каждого набора:
results_stemmed_count 0.7492
results_lemmatized_count 0.784
results_stemmed_tfidf 0.7532
results_lemmatized_tfidf 0.7944
results_stemmed_count для набора (1, 1)
максимальный скор: 0.7492
[[0.5652 0.5652 0.5652 0.5652 0.5652 0.5652 0.5652]
 [0.6672 0.6672 0.6672 0.6672 0.6672 0.6672 0.6672]
 [0.722  0.722  0.722  0.722  0.722  0.722  0.722 ]
 [0.7284 0.7284 0.7284 0.7284 0.7284 0.7284 0.7284]
 [0.7396 0.7396 0.7396 0.7396 0.7396 0.7396 0.7396]
 [0.7492 0.7492 0.7492 0.7492 0.7492 0.7492 0.7492]]
results_lemmatized_count для набора (1, 1)
максимальный скор: 0.7812
[[0.5604 0.5604 0.5604 0.5604 0.5604 0.5604 0.5604]
 [0.6692 0.6692 0.6692 0.6692 0.6692 0.6692 0.6692]
 [0.7492 0.7492 0.7492 0.7492 0.7492 0.7492 0.7492]
 [0.7652 0.7652 0.7652 0.7652 0.7652 0.7652 0.7652]
 [0.778  0.778  0.778  0.778  0.778  0.778  0.778 ]
 [0.7812 0.7812 0.7812 0.7812 0.7812 0.7812 0.7812]]
results_stemmed_tfidf для набора (1, 1)
максимальный скор: 0.7532
[[0.5632 0.5632 0.

esults_stemmed_count 0.7492

results_lemmatized_count 0.784

results_stemmed_tfidf 0.7532

results_lemmatized_tfidf 0.7944


Анализируя, можно прийти к выводу, что лучший эффект достигается при лемматизациии, нежели стемминге.
Также можно заметить, что tfidf обеспечивает наибольший скор.

results_stemmed_count для набора (1, 1)
максимальный скор: 0.7492



По горизонтали изменение изменение max_dfs, по вертикали - изменение max_features

In [79]:
max_dfs

[0.5, 0.6, 0.7, 0.8, 0.9, 0.95, 0.99]

In [81]:
max_featureses

[10, 100, 1000, 2000, 5000, 10000]

In [77]:
results_stemmed_count[:,:,0]

array([[0.5652, 0.5652, 0.5652, 0.5652, 0.5652, 0.5652, 0.5652],
       [0.6672, 0.6672, 0.6672, 0.6672, 0.6672, 0.6672, 0.6672],
       [0.722 , 0.722 , 0.722 , 0.722 , 0.722 , 0.722 , 0.722 ],
       [0.7284, 0.7284, 0.7284, 0.7284, 0.7284, 0.7284, 0.7284],
       [0.7396, 0.7396, 0.7396, 0.7396, 0.7396, 0.7396, 0.7396],
       [0.7492, 0.7492, 0.7492, 0.7492, 0.7492, 0.7492, 0.7492]])

Можно заметить, что изменение max_df для этого набора вносит меньший вклад (изменение по горизонтали), чем изменение max_features

results_lemmatized_count для набора (1, 2)
максимальный скор: 0.784

In [82]:
results_lemmatized_count[:,:,1]

array([[0.5604, 0.5604, 0.5604, 0.5604, 0.5604, 0.5604, 0.5604],
       [0.668 , 0.668 , 0.668 , 0.668 , 0.668 , 0.668 , 0.668 ],
       [0.7452, 0.7452, 0.7452, 0.7452, 0.7452, 0.7452, 0.7452],
       [0.7616, 0.7616, 0.7616, 0.7616, 0.7616, 0.7616, 0.7616],
       [0.778 , 0.778 , 0.778 , 0.778 , 0.778 , 0.778 , 0.778 ],
       [0.784 , 0.784 , 0.784 , 0.784 , 0.784 , 0.784 , 0.784 ]])

Для лемматизации аналогичная картина

results_stemmed_tfidf для набора (1, 1)
максимальный скор: 0.7532


In [84]:
results_stemmed_tfidf[:,:,0]

array([[0.5632, 0.5632, 0.5632, 0.5632, 0.5632, 0.5632, 0.5632],
       [0.6692, 0.6692, 0.6692, 0.6692, 0.6692, 0.6692, 0.6692],
       [0.7208, 0.7208, 0.7208, 0.7208, 0.7208, 0.7208, 0.7208],
       [0.7396, 0.7396, 0.7396, 0.7396, 0.7396, 0.7396, 0.7396],
       [0.7404, 0.7404, 0.7404, 0.7404, 0.7404, 0.7404, 0.7404],
       [0.7532, 0.7532, 0.7532, 0.7532, 0.7532, 0.7532, 0.7532]])

Для стемминга и tf-idf тоже самое.

results_lemmatized_tfidf для набора (1, 1)
максимальный скор: 0.7944


In [85]:
results_lemmatized_tfidf[:,:,0]

array([[0.562 , 0.562 , 0.562 , 0.562 , 0.562 , 0.562 , 0.562 ],
       [0.6696, 0.6696, 0.6696, 0.6696, 0.6696, 0.6696, 0.6696],
       [0.754 , 0.754 , 0.754 , 0.754 , 0.754 , 0.754 , 0.754 ],
       [0.7704, 0.7704, 0.7704, 0.7704, 0.7704, 0.7704, 0.7704],
       [0.784 , 0.784 , 0.784 , 0.784 , 0.784 , 0.784 , 0.784 ],
       [0.7944, 0.7944, 0.7944, 0.7944, 0.7944, 0.7944, 0.7944]])

Для лемматизации и tf-idf ничего не меняется.

Таким образом, можно прийти к выводу, что изменение количества признаков взосит больший вклад в увеличение итоговой точности, чем изменение порога (целенаправлено выбран порог от 0,5 до 0,9 с шагом 0,1 и два значения 0,95 и 0,99).

Также можно заметить, что увеличение количества признаков от 2000 даёт низкий прирост скора, но сильнее усложняет расчёт.

Интересное наблюдение - максимальный скор показали варианты с n-граммами размерностью 1 и 2 (в меньшей степени). Варианты с биграммами и триграммами дали меньший скор, но можно предположить, что трубется большее число фичей.

In [89]:
for ngram_range in ngram_ranges:
  lemmatized_tfidf = TfidfVectorizer(ngram_range=ngram_range, 
                              analyzer='word', 
                              binary=False, 
                              tokenizer=str.split,
                              max_df=0.9,
                              max_features=1000000,
                              stop_words='english')
  lemmatized_tfidf = lemmatized_tfidf.fit(tweet_lemmatized_doc)
  xtrain_count =  lemmatized_tfidf.transform(train_x)
  xvalid_count =  lemmatized_tfidf.transform(valid_x)
  classifier = linear_model.LogisticRegression()
  classifier.fit(xtrain_count, train_y)
  predictions = classifier.predict(xvalid_count)
  print(accuracy_score(valid_y, predictions), ngram_range)

0.7952 (1, 1)
0.7936 (1, 2)
0.7936 (1, 3)
0.5996 (2, 2)
0.5996 (2, 3)
0.5064 (3, 3)


При максимальном числе фичей 1 000 000 в текущей конфигурации также "побеждает" ngram_range=(1,1)

Как применить РСА я, к сожалению, не понял, нужны подсказки.